### PR TITLE
multi-field sorting

### LINKED ISSUE #1141: Sorting unread items first
Is there a way to sort the items of a feed by date with unread items first? What happens right now is that items are simply sorted by date, and if a new item is added with an older date it could end up at the end of the list.

### PR DESCRIPTION
**Pull request by srijita2506**
Fixes #1141

This PR updates article-sort-order so it can use more than one sort field.
article-sort-order "date-asc,unread"-This will sort by date first with newest articles first and then put the unread articles before the read articles.
Can work on multiple use-cases now
Article sort parsing accepts multiple sort fields.
added unread as sort field
worked on article sorting to sort in sequence using stable sort.
changed documentation for new config
added test for multi field sorting
added test for unread and date sorting

### CODE REVIEW

**Comment by srijita2506** [comment]
@Minoru,Rightnow if there are two fields to be sorted upon then it first sorts through the first field and if there is a tie then it sorts through the second field, so in the command if user puts ---set article-sort-order unread-desc,date-asc---then it would sort with keeping unread articles at top and if there are two unreads or reads then it will give preference to the newer one

File: test/configcontainer.cpp [hunk 1]
```diff
@@ -564,124 +564,127 @@ TEST_CASE(
...
+		return sort_strategy.rules.at(index);
+	};
 
 	SECTION("title") {
 		cfg.set_configvalue("article-sort-order", "title");
 		sort_strategy = cfg.get_article_sort_strategy();
-		REQUIRE(sort_strategy.sm == ArtSortMethod::TITLE);
-		REQUIRE(sort_strategy.sd == SortDirection::ASC);
+		REQUIRE(rule(0).sm == ArtSortMethod::TITLE);
+		REQUIRE(rule(0).sd == SortDirection::ASC);
```
Comment (Minoru) [inline test/configcontainer.cpp:575]: Please check the length of `sort_strategy.rules` like you do in tests below.

File: test/configcontainer.cpp
Comment (Minoru) [inline test/configcontainer.cpp:1]: There should be tests for parsing multiple rules like "date,title-desc" etc., along with corner cases like "foo,bar-baz" etc.

File: test/rssfeed.cpp [hunk 2]
```diff
@@ -12,6 +12,14 @@
 
 using namespace newsboat;
 
+static ArticleSortStrategy make_article_sort_strategy(
```
Comment (Minoru) [inline test/rssfeed.cpp:15]: We prefer anonymous namespaces.

File: test/rssfeed.cpp [hunk 3]
```diff
@@ -209,6 +217,99 @@ TEST_CASE("RssFeed::sort() correctly sorts articles", "[RssFeed]")
 		REQUIRE(articles[3]->pubDate_timestamp() == 23);
 		REQUIRE(articles[4]->pubDate_timestamp() == 7);
 	}
+
+	SECTION("date-asc,unread") {
```
Comment (Minoru) [inline test/rssfeed.cpp]: TK: default order for everything except "date" is "asc", but the `ss` below is initialised with `SortDirection::DESC`. Please update either the code or the section name.

File: test/rssfeed.cpp [hunk 4]
```diff
@@ -209,6 +217,99 @@ TEST_CASE("RssFeed::sort() correctly sorts articles", "[RssFeed]")
...
+		REQUIRE(articles[0]->pubDate_timestamp() == 7);
+		REQUIRE(articles[1]->unread());
+		REQUIRE(articles[1]->pubDate_timestamp() == 23);
+		REQUIRE(articles[2]->unread());
+		REQUIRE(articles[2]->pubDate_timestamp() == 42);
+		REQUIRE_FALSE(articles[3]->unread());
+		REQUIRE(articles[3]->pubDate_timestamp() == 69);
+		REQUIRE_FALSE(articles[4]->unread());
+		REQUIRE(articles[4]->pubDate_timestamp() == 93);
+	}
```
Comment (Minoru) [inline test/rssfeed.cpp:281]: These tests aren't very descriptive because the dates are all different and sorting by unread does nothing. Perhaps make some of the dates the same to show how sorting by unread affects the result?

File: test/rssfeed.cpp [hunk 5]
```diff
@@ -209,6 +217,99 @@ TEST_CASE("RssFeed::sort() correctly sorts articles", "[RssFeed]")
...
+		REQUIRE(articles[0]->pubDate_timestamp() == 42);
+		REQUIRE(articles[1]->unread());
+		REQUIRE(articles[1]->pubDate_timestamp() == 23);
+		REQUIRE_FALSE(articles[2]->unread());
+		REQUIRE(articles[2]->pubDate_timestamp() == 93);
+		REQUIRE_FALSE(articles[3]->unread());
+		REQUIRE(articles[3]->pubDate_timestamp() == 69);
+		REQUIRE_FALSE(articles[4]->unread());
+		REQUIRE(articles[4]->pubDate_timestamp() == 7);
+	}
```
Comment (Minoru) [inline test/rssfeed.cpp:312]: These tests have the same setup. Combine them into a section to avoid code repetition:

```
SECTION("by date and unread") {
    auto articles = f.items();
    articles[0]->set_pubdate(93);
    ...

    SECTION("date-asc,unread") {
        auto ss = make_article_sort_strategy(...);
        ...
    }

    SECTION("date-desc,unread") {
    ...
```

File: src/configcontainer.cpp [hunk 6]
```diff
@@ -611,37 +611,59 @@ FeedSortStrategy ConfigContainer::get_feed_sort_strategy() const
...
-		ss.sm = ArtSortMethod::DATE;
-		ss.sd = SortDirection::DESC;
-		if (methods.size() > 1 && methods[1] == "asc") {
-			ss.sd = SortDirection::ASC;
-		}
-	} else { // all other sort methods are ascending by default
-		ss.sd = SortDirection::ASC;
-		if (methods.size() > 1 && methods[1] == "desc") {
-			ss.sd = SortDirection::DESC;
+	ss.rules.clear();
```
Comment (Minoru) [inline src/configcontainer.cpp:614]: Nit: what's the point of clearing the structure that was just created?

File: src/configcontainer.cpp [hunk 7]
```diff
@@ -611,37 +611,59 @@ FeedSortStrategy ConfigContainer::get_feed_sort_strategy() const
...
+
+	const auto settings =
+		utils::tokenize(get_configvalue("article-sort-order"), ",");
+	for (const auto& setting : settings) {
+		const auto methods = utils::tokenize(setting, "-");
+		if (methods.empty()) {
+			continue;
 		}
```
Comment (Minoru) [inline src/configcontainer.cpp:622]: Need a test for that.

File: src/configcontainer.cpp [hunk 8]
```diff
@@ -611,37 +611,59 @@ FeedSortStrategy ConfigContainer::get_feed_sort_strategy() const
...
+				rule.sd = SortDirection::DESC;
+			}
+		}
+
+		ss.rules.push_back(rule);
+	}
+
+	if (ss.rules.empty()) {
+		ss.rules.push_back(ArticleSortRule{});
 	}
```
Comment (Minoru) [inline src/configcontainer.cpp:667]: Need a test for this too.

File: src/configcontainer.cpp [hunk 9]
```diff
@@ -611,37 +611,59 @@ FeedSortStrategy ConfigContainer::get_feed_sort_strategy() const
...
+
+		if (!known_method) {
+			continue;
 		}
+
+		if (rule.sm == ArtSortMethod::DATE || rule.sm == ArtSortMethod::UNREAD) {
+			rule.sd = SortDirection::DESC;
+			if (methods.size() > 1 && methods[1] == "asc") {
+				rule.sd = SortDirection::ASC;
+			}
```
Comment (Minoru) [inline src/configcontainer.cpp:654]: Why does "unread" use "desc" by default?

**Review by Minoru — CHANGES_REQUESTED** [review]
It's a good start!

In addition to my other commends, this PR also needs implementation for `feed-sort-order` and the updated documentation for both.

Please make separate commits for logically separate changes (google "atomic commits"). In this PR, "unread" and multi-key sorting are totally separate, and it'd be easier to review two small-ish commits instead of one large diff.

Also, please run a code formatter on these changes (we have a Dockerfile with the specific version we're using).

**Comment by srijita2506** [comment]
@Minoru Thankyou for your review.
I will fix those issues.
It would just take some time as I am still in my university.
Thankyou once again I learnt a lot by working on this issue.
